### DISCUSSION TITLE
Allow inline variable declaration with "as" operator

### DISCUSSION THREAD

**Discussion by jonathanmarston** [dotnet/csharplang#4843] [Language Ideas]
## Proposal
This proposal would introduce the ability to declare a variable when using the ```as``` operator, similar to inline declaration with the ```is``` pattern. 

## Motivation
Many times it is desirable to capture the result of an expression as a variable to reuse later as part of a larger expression. This can be simply for the sake of brevity and convenience, or even to avoid additional overhead of re-parsing/processing data and redundant allocations without breaking out into multiple statements.

There is precedent for a feature like this in both the ```is``` operator, and inline declaration of variables for ```out``` parameters. In fact, there is currently a strange gap in the language that allows inline declaration of variables to hold the value of ```out``` parameters, but no way to declare inline variables to hold return values. 

There is also the ```let``` clause in LINQ, which is somewhat similar in concept, but only usable within the context of a query, and does not cover the same set of use cases.

Other proposals like expression blocks (#3086) can partially allow for this, but the syntax is not as clean for many use cases since block expressions would still require breaking the expression up into multiple statement (even if contained within a block expression) and would require a typical "right-to-left" assignment, which breaks the "left-to-right" flow of expressions. 

## Design
Just as inline declarations are allowed with the ```is``` pattern, the ```as``` operator would be extended to allow declaring variables inline, the main difference being that the evaluated result of the expression is the value itself (or null, when a reference type conversion fails), and not a ```bool```. 

When the type of the left side of the expression and the declared type exactly match, the safe conversion would not be emitted, and only the inline variable declaration and assignment would exist in the compiled code.

The ```as``` operator  would also be extended to work with value types, but only when the declared (or inferred) type exactly matches the type of the expression on the left. No conversion would ever be performed on value types, whether implicit or explicit, and as such, would only be useful for inline variable declaration in that context.

Type inference would also be allowed through the use of ```var```.

## Examples
Using inline variable declaration with ```as```, this:
```csharp
// "path" is a string
string filename = path.LastIndexOf('/') as int index > -1 ? path.Substring(index + 1) : path; 
```

Is equivalent to this:
```csharp
// "path" is a string
int index;
string filename = (index = path.LastIndexOf('/')) > -1 ? path.Substring(index + 1) : path; 
```

The ```as``` operator would work just as before for reference types, but now it can also be used with value types as long as no conversion is taking place:

```csharp
// "path" is a string
string filename = path.LastIndexOf('/') as double index > -1.0 ? path.Substring((int)index + 1) : path; // Compiler error. Cannot convert 'int' to 'double'
string filename = ((double)path.LastIndexOf('/')) as double index > -1.0 ? path.Substring((int)index + 1) : path; // Allowed, but an admittedly contrived example.
int one = 1 as int; // Allowed, but does nothing.


// "path" is an object
string filename = (path as string pathString)?.LastIndexOf('/') as int index > -1 ? pathString.Substring(index + 1) : pathString; // Compiler error. Cannot convert 'int?' to 'int'
string filename = (path as string pathString)?.LastIndexOf('/') as int? index > -1 ? pathString.Substring((int)index + 1) : pathString; // Allowed
```

Although the previous example would work, a better way to do it would actually be to use inline declaration with both ```is``` and ```as```:
```csharp
// "path" is an object
string filename = path is string pathString && pathString.LastIndexOf('/') as int index > -1 ? pathString.Substring(index + 1) : pathString;
```

Type inference would also be allowed through ```var```:
```csharp
// "path" is a string
string filename = path.LastIndexOf('/') as var index > -1 ? path.Substring(index + 1) : path; // Allowed
```
## Arguments
This would essentially expand the ```as``` operator to cover 2 use cases, which could be used both independently and concurrently:
1. A safe conversion operator (existing)
2. A way to capture a value mid-expression through inline declaration (new)

Overloading an existing operator in this way could be seen as confusing to some, but it does fit in well with existing syntax around the (similar, but different) ```is``` pattern matching operator, and therefore seems like a mostly natural extension of the existing language.

**Comment by FaustVX** [comment]
Pattern matching with `is` works just like that.
```csharp
string filename = path.LastIndexOf('/') is int index and > -1 ? path[(index + 1)..] : path;
```
[Sharplab.io](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD4DcLIAQAOECAFngLx4BEAZgPYA2AJgKbID0NUDLVA3AFgAUAAEAjAAY8XHjAhgWFQsRIA6ADIQAzggCSMVgA8A8jQAUAcnYWAlHihb7MBE6N4IBvAD48AWjF4APzKpADaZrBuANR4YjaqqgC6eCAhJIKiYqriAJxmMixyCjZ8QA==)

_Notice the `and` keyword_
